libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/idp/IDPInititatedLoginReceiver.java

```diff
@@ -0,0 +1,102 @@
+ * All rights reserved.
+ * Redistribution and use of this software in source and binary forms, with or
+ * without modification, are permitted provided that the following conditions
+ * are met:
+ * - Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ * - Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * - Neither the name of salesforce.com, inc. nor the names of its contributors
+ * may be used to endorse or promote products derived from this software without
+ * specific prior written permission of salesforce.com, inc.
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.salesforce.androidsdk.auth.idp;
+
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.os.Bundle;
+import android.text.TextUtils;
+
+import com.salesforce.androidsdk.accounts.UserAccount;
+import com.salesforce.androidsdk.app.SalesforceSDKManager;
+
+/**
+ * This class receives a request to login from an IDP app and triggers the
+ * IDP login flow from within the SP app based on the 'user_hint' passed in.
+ *
+ * @author bhariharan
+ */
+public class IDPInititatedLoginReceiver extends BroadcastReceiver {
+
+    public static final String IDP_LOGIN_REQUEST_ACTION = "com.salesforce.IDP_LOGIN_REQUEST";
+    public static final String USER_HINT_KEY = "user_hint";
+    public static final String IDP_INIT_LOGIN_KEY = "idp_init_login";
+    private static final String COLON = ":";
+
+    private String userHint;
+
+    @Override
+    public void onReceive(Context context, Intent intent) {
+        if (intent != null) {
+            if (IDP_LOGIN_REQUEST_ACTION.equals(intent.getAction())) {
+                final Bundle extras = intent.getExtras();
+                if (extras != null) {
+                    userHint = extras.getString(USER_HINT_KEY);
+                }
+
+                // Launches login flow if the user doesn't already exist on the SP app.
+                if (!doesUserExist()) {
```

Comment:
Checks if the user passed in already exists on the SP app. If it does, there's nothing to do.

Reply:
It feels like something is missing. The IDP should what we want the SP app to do once we have the desired user.
It would be like the returl on the server side (when you have to go through login in the middle of a flow but and want the flow to resume).

Reply:
@trooper2013 what do you do on the iOS side?

Reply:
Yeah, makes sense. I was going to enhance this in the next PR to take in an optional component/intent from the IDP that can be launched if present. Right now it ends back on the IDP since `LoginActivity` will be popped off the stack once logged in.

Reply:
Sounds good.